File: src/main/python/binarysearch/나무자르기2805.py
import sys
N, M = map(int, sys.stdin.readline().split())
forest = list(map(int, sys.stdin.readline().split()))
s = 0
e = max(forest)

while s <= e : # 교차되기 전까지!
  mid = (s+e)//2

  # 얼마나 나무를 채취했나요?
  wood = sum(tree - mid for tree in forest if tree > mid)

  # 업인가요? 다운인가요?

  if wood >= M : #최소를 만족했다!
    s = mid +1
  else:
    e = mid -1

print(e)

File: src/main/java/org/snust/binarysearch/나무자르기2805.java
package org.snust.binarysearch;

import java.io.*;
import java.util.*;

public class 나무자르기2805 {
  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());

    int N = Integer.parseInt(st.nextToken());
    int M = Integer.parseInt(st.nextToken());

    int[] forest = new int[N];
    st = new StringTokenizer(br.readLine());
    int maxHeight = 0;

    for(int i = 0; i < N; i++) {
      forest[i] = Integer.parseInt(st.nextToken());
      maxHeight = Math.max(maxHeight, forest[i]);
    }

    long s = 0;
    long e = maxHeight;

    while(s <= e) {
      long mid = (s + e) / 2;
      long wood = 0;

      for(int tree : forest) {
        if(tree > mid) {
          wood += tree - mid;
        }
      }

      if(wood >= M) {
        s = mid + 1;
      } else {
        e = mid - 1;
      }
    }

    System.out.println(e);
  }
}
